src/ray/gcs/subscribe_executor.cc

```diff
@@ -0,0 +1,106 @@
+#include "ray/gcs/subscribe_executor.h"
+
+namespace ray {
+
+namespace gcs {
+
+template <typename ID, typename Data, typename Table>
+Status SubscribeExecutor<ID, Data, Table>::AsyncSubscribe(
+    const ClientID &client_id, const SubscribeCallback<ID, Data> &subscribe,
+    const StatusCallback &done) {
+  // TODO(micafan) Optimize the lock when necessary.
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  if (registered_) {
+    return Status::OK();
+  }
+
+  auto on_subscribe = [this, subscribe](RedisGcsClient *client, const ID &id,
+                                        const std::vector<Data> &result) {
+    if (result.empty()) {
+      return;
+    }
+
+    SubscribeCallback<ID, Data> callback = nullptr;
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      const auto it = id_to_request_map_.find(id);
+      if (it != id_to_request_map_.end()) {
+        callback = it->second.subscribe_;
+      }
+    }
+    if (callback != nullptr) {
+      callback(id, result.back());
+    } else if (subscribe != nullptr) {
+      subscribe(id, result.back());
+    }
```

Comment:
If user subscribes the whole table and a specific item, I think both callbacks should be invoked?

Reply:
If someone subscribes the whole table, why subscribe to a specific item again? It's not logical I think.

Reply:
Also, I should check either subscribe the whole table or specifics items, not both.

Reply:
I agree with you that this should not happen in common cases. I think either one of the below approach should be fine:
- reject the case when user tries to subscribe both whole table and a single item, fail the second subscription in this case;
- allow this case to happen, then we should invoke both callbacks.

it seems simpler just to reject it.

Reply:
Acctually invoke both callbacks is a little simpler. How about adopt this simple solution?